Improve const-correctness of JIT / fix bug where ONNX verbose=1 export says the wrong thing

ezyang opened the pull request:
This started off as a minor fix based on Adam's question, "why is printing
a graph not const" and snowballed into a giant yak shaving exercise.

- The Graph and Node APIs now uniformly enforce deep constness; e.g., if you
  get a const Node* or const Graph*, it is not possible to get a non-const
  Node*/Graph* somewhere else in the graph (even though the member variables
  of these are non-const.  Hooray for private access specifier.)

- A big pile of functions got const versions, most notably the printing
  functions, and functions for accessing inputs().

- REALLY IMPORTANT, BC-BREAKING CHANGE: inputs() now returns a COPY of the
  inputs, rather than a reference to the underlying.  I was forced to do this
  because there is no way to portably turn a std::vector<Node*> into a
  std::vector<const Node*>, which is necessary to provide a const-correct
  version of inputs() that enforces deep const-correctness.  I then justified
  this choice to myself with the observation that outputs() returned a
  copy (by necessity), so this makes the API more uniform.

  But making this change uncovered two very subtle bugs:

    1. If you change functions from returning a reference to returning a copy,
       the idiom node->inputs().begin() is no longer valid, because the memory
       the iterator points to immediately becomes invalid.  THIS SUCKS.
       Honestly, we should add a lint rule rejecting calling begin()/end() on
       temporaries because this is very dangerous.  To excise this pattern from
       the codebase, I added begin() and end() methods to Graph, so that we got
       rid of the graph->nodes().begin() idiom, which happens to be sound,
       despite not returning a reference, because graph_node_list is a
       non-owning reference.

    2. pybind11 doesn't handle std::vector<Node*> cast out of the box.
       Fortunately, I found a simple fix in the GitHub issues tracker
       that involved adding an extra type converter.  And yes, this
       does mean that outputs() in Python never worked correctly.

- New const_graph_node_list, which is a graph_node_list that gives you const
  Node*

There are some more miscellaneous improvements:

- Applied CR comment fixes on export.cpp; using replaceInput, and renaming
  variables for clarity.

- assertValidInput helper method added, and applied to replaceInput

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Improve const-correctness of JIT.

This started off as a minor fix based on Adam's question, "why is printing
a graph not const" and snowballed into a giant yak shaving exercise.

- The Graph and Node APIs now uniformly enforce deep constness; e.g., if you
  get a const Node* or const Graph*, it is not possible to get a non-const
  Node*/Graph* somewhere else in the graph (even though the member variables
  of these are non-const.  Hooray for private access specifier.)

- A big pile of functions got const versions, most notably the printing
  functions, and functions for accessing inputs().

- REALLY IMPORTANT, BC-BREAKING CHANGE: inputs() now returns a COPY of the
  inputs, rather than a reference to the underlying.  I was forced to do this
  because there is no way to portably turn a std::vector<Node*> into a
  std::vector<const Node*>, which is necessary to provide a const-correct
  version of inputs() that enforces deep const-correctness.  I then justified
  this choice to myself with the observation that outputs() returned a
  copy (by necessity), so this makes the API more uniform.

  But making this change uncovered two very subtle bugs:

    1. If you change functions from returning a reference to returning a copy,
       the idiom node->inputs().begin() is no longer valid, because the memory
       the iterator points to immediately becomes invalid.  THIS SUCKS.
       Honestly, we should add a lint rule rejecting calling begin()/end() on
       temporaries because this is very dangerous.  To excise this pattern from
       the codebase, I added begin() and end() methods to Graph, so that we got
       rid of the graph->nodes().begin() idiom, which happens to be sound,
       despite not returning a reference, because graph_node_list is a
       non-owning reference.

    2. pybind11 doesn't handle std::vector<Node*> cast out of the box.
       Fortunately, I found a simple fix in the GitHub issues tracker
       that involved adding an extra type converter.  And yes, this
       does mean that outputs() in Python never worked correctly.

- New const_graph_node_list, which is a graph_node_list that gives you const
  Node*

There are some more miscellaneous improvements:

- Applied CR comment fixes on export.cpp; using replaceInput, and renaming
  variables for clarity.

- assertValidInput helper method added, and applied to replaceInput

- Use an explicit function to print THPObjectPtr, otherwise we get
  the wrong overload.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

zdevito commented:
Haven't had time to look at this fully yet, but its not clear if we should be copying the inputs list every time just because of const correctness. Instead can we make both input and output not make copies? For instance, can we return ArrayRef<Node*> or ArrayRef<const Node*> for both input and output?  You can construct a single-entry ArrayRef for the `outputs().size() == 1` whose begin  is just the node itself. I think the only gotcha is making sure the python conversions of ArrayRef work.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -167,20 +171,40 @@ struct Node : public Attributes<Node> {
+    for(auto & u : uses())
+      r.push_back(u.user);
+    return r;
+  }
```
eh having two implementations for everything is really annoying...

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -423,6 +444,17 @@ struct Node : public Attributes<Node> {
   }
 
+  // Enforce the select invariant on inputs
+  void assertValidInput(const Node *node) const {
```
maybe prefix it with underscore or at least mark private?

ezyang replied:
Well, it's already private, but I can add an underscore ;)

ezyang replied:
Actually, if I add an underscore, I'm going to want to apply this consistently to all private methods

apaszke reviewed (commented):
Looks good. I hate that C++ doesn't let you do this and you have to reimplement everything twice...

ezyang commented:
> Its not clear if we should be copying the inputs list every time just because of const correctness. Instead can we make both input and output not make copies? For instance, can we return ArrayRef<Node*> or ArrayRef<const Node*> for both input and output?

I considered doing this; it certainly avoids the problem where `std::vector<Node*>` and `std::vector<const Node*>` are not interconvertible. The primary problem is that while you now get to see updates to the underlying vector, if it *resizes*, your array reference will continue to have a stale size. If we're OK with this, I can switch it over.

apaszke commented:
I don't mind having a copy, since it's what we're doing without `outputs` anyway

zdevito commented:
I think we should remove both copies. The only reason I had outputs copy was because there wasn't a great way to avoid it without writing more pybind wrappers. Now that we (apparently) have to add pybind wrappers anyway, we might as well make both not copy. I am OK with inputs() and outputs() being views ArrayRef's make the fact that it is a view clear in the type signature.

ezyang commented:
@zdevito Actually, I tried making them return not-copies, but this doesn't work. Look at the implementation of `outputs()`:

```
  std::vector<const Node*> outputs() const {
    if(!hasMultipleOutputs())
      return { this };
    std::vector<const Node*> r;
    r.reserve(uses().size());
    for(auto & u : uses())
      r.push_back(u.user);
    return r;
  }
```

Due to the representation of uses, we MUST materialize a new vector, which means it has to be a copy. I suppose we could rewrite the representation of uses so that you get a vector when everything is select nodes, but this feels like the tail wagging the dog.

zdevito commented:
I do think inputs should make a copy. It is very easy to use it in a loop thinking you have a reference to the inputs, and It is surprising behavior. Apparently we need to do this for outputs, which I also think is surprising behavior, but I'd rather fix outputs rather than regress inputs.

ezyang committed:
Move ONNX broadcast fusion into separate ONNX pass, fixes verbose printing.

This breaks a lot of the onnx-pytorch tests because the abstraction
barriers are not respected.  I'll spin up a patch for that separately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
OK, let's just make inputs() return an ArrayRef, outputs() return a copy, and document it.

ezyang committed:
Make inputs/outputs return an ArrayRef.

Some knock on effects:

- at() is not supported on ArrayRef.  I fixed this by adding a new
  overload for input() to access a specific input.  I also filed
  https://github.com/zdevito/ATen/pull/152

- Need new overloads for fmap/filter, because template deduction won't
  attempt an implicit constructor in attempt to match the argument.

- New overload in ir.cpp for printing ArrayRef.

- When we pybind11 an ArrayRef, we convert it into an iterator.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
